content/Hash functions in daily life I - Martin Rubi.md

```diff
@@ -0,0 +1,172 @@
+                  | . | . | . | . | . | . |
+                  -------------------------
+
+
+The first empty space was either an empty box or a partially filled box with
+enough space for another bag.
+
+If the grid had no space left the bags were stacked aside.
+
+                  -------------------------
+                  | . | . | . | . | . | . |
+                  -------------------------
+                  | . | . | . | . | . | . |
+                  -------------------------
+                  | . | . | . | . | . | . |
+                  -------------------------   ...
+
+
+To find the bag with a matching order number the cashier performed a lookup upon
+all the bags in the grid.
+
+That's a linear lookup. It's O(N), meaning that, let N be the number of
+laundry bags stored in the grid, the lookup would take an average of N comparisons
+to find the searched one.
+
+Actually it would take an average of N/2 comparisons but the order of magnitude
+between N and N/2 is the same and it´s said to be O(N) for simplicity.
+
+To find a given bag could take less or more than N/2 but in average, given an evenly
+distributed set of bags, it would take N/2 comparisons.
+
+## Meet a hash function
+
+Each laundry order number was a sequencial integer number. The number was created
+at the time of leaving the laundry in the store and was printed in the ticket.
+
+The order number had around 10 digits digits.
+
+The last digit of each order number was a number between 0 and 9.
+
+That last digit could be used to optimize the lookup.
+
+To make the optimization the grid might be divided into different boxes.
+
+Some boxes would have a label assigned with a digit between 0 and 9.
+
+For example
+
+          -------------------------
+          |   |   |   |   |   |   |
+          -------------------------
+          | 7 | 8 | 9 | 0 |   |   |
+          -------------------------
+          | 1 | 2 | 3 | 4 | 5 | 6 |
+          -------------------------
+
+At the time of storing the laundry the cashier would look at the order number and
+take its last digit only. If the box corresponding to the last digit had empty
+space the bag would go into that box.
+
+If the box had not enough space for a new bag the bag would go to the first box
+with space among the unlabeled boxes.
+
+
+The lookup of a bag would apply the same algorithm. First she would look for the
+bag in the box corresponding to the last digit of the order. It the bag wasn't
+there she would look for it in the unlabeled boxes.
+
+That lookup is still O(N) in its worst case but in the average case it's O(1),
+meaning that, in average, the lookup would take only a constant number of comparisons.
+
+O(1) doesn't mean that it would take exactly one comparison. It means that the
+number of comparisons is a constant number and it does not depend on the total
+number of bags. In this case the constant number would be the number of bags that
+could fit in a single box.
```

Comment:
```suggestion
O(1) doesn't mean that it would take exactly one comparison. It means that the
number of comparisons is a constant number and it does not depend on the total
number of bags. In this case, the constant number would be the number of bags that
could fit in a single box.
```